x-pack/legacy/plugins/integrations_manager/server/integrations/install.ts

```diff
@@ -0,0 +1,210 @@
+  const toSave = await installAssets({ savedObjectsClient, pkgkey, asset, callCluster });
+
+  if (toSave.length) {
+    // saved those references in the integration manager's state object
+    const saved = await saveInstallationReferences({ savedObjectsClient, pkgkey, toSave });
+    return saved;
+  }
+
+  return [];
+}
+
+// the function which how to install each of the various asset types
+// TODO: make it an exhaustive list
+// e.g. switch statement with cases for each enum key returning `never` for default case
+export async function installAssets(options: {
+  savedObjectsClient: SavedObjectsClientContract;
+  pkgkey: string;
+  asset: AssetType;
+  callCluster: CallESAsCurrentUser;
+}) {
+  const { savedObjectsClient, pkgkey, asset, callCluster } = options;
+  if (assetUsesObjects(asset)) {
+    const references = await installObjects({ savedObjectsClient, pkgkey, asset });
+    return references;
+  }
+  if (asset === ASSET_TYPE_INGEST_PIPELINE) {
+    const references = await installPipelines({ callCluster, pkgkey });
+    return references;
+  }
+
+  return [];
+}
+
+export async function saveInstallationReferences(options: {
+  savedObjectsClient: SavedObjectsClientContract;
+  pkgkey: string;
+  toSave: AssetReference[];
+}) {
+  const { savedObjectsClient, pkgkey, toSave } = options;
+  const savedObject = await getInstallationObject({ savedObjectsClient, pkgkey });
+  const savedRefs = savedObject && savedObject.attributes.installed;
+  const mergeRefsReducer = (current: AssetReference[], pending: AssetReference) => {
+    const hasRef = current.find(c => c.id === pending.id && c.type === pending.type);
+    if (!hasRef) current.push(pending);
+    return current;
+  };
+
+  const toInstall = toSave.reduce(mergeRefsReducer, savedRefs || []);
+  const results = await savedObjectsClient.create<InstallationAttributes>(
+    SAVED_OBJECT_TYPE,
+    { installed: toInstall },
+    { id: pkgkey, overwrite: true }
+  );
+
+  return results;
+}
+
+async function installObjects({
+  savedObjectsClient,
+  pkgkey,
+  asset,
+}: {
+  savedObjectsClient: SavedObjectsClientContract;
+  pkgkey: string;
+  asset: AssetType;
+}): Promise<AssetReference[]> {
+  const filter = (entry: Registry.ArchiveEntry) => asset === Registry.pathParts(entry.path).type;
+  const toBeSavedObjects = await getObjects(pkgkey, filter);
+  const createResults = await savedObjectsClient.bulkCreate<InstallationAttributes>(
+    toBeSavedObjects,
+    { overwrite: true }
+  );
+  const createdObjects = createResults.saved_objects;
+  const installed = createdObjects.map(({ id, type }) => ({ id, type }));
+
+  return installed;
+}
+
+async function installPipelines({
+  callCluster,
+  pkgkey,
+}: {
+  callCluster: CallESAsCurrentUser;
+  pkgkey: string;
+}) {
+  const isPipeline = ({ path }: Registry.ArchiveEntry) =>
+    Registry.pathParts(path).type === ASSET_TYPE_INGEST_PIPELINE;
+  const paths = await Registry.getArchiveInfo(`${pkgkey}.tar.gz`, isPipeline);
```

Comment:
Is it safe to construct the filename from `pkgkey`?

Reply:
It is, but it’s an implementation detail that shouldn’t be in this file. I’ll export functions from registry to create it.

Reply:
> I’ll export functions from registry to create it.

I thought more places were using this. Went with a simpler change.  https://github.com/elastic/kibana/pull/41073/commits/3e8db8500cc1b7b7f1c4f1424080cb559cc51a7a